lib/web_ui/lib/src/engine/host_node.dart

```diff
@@ -0,0 +1,161 @@
+
+part of engine;
+
+/// The interface required to host a flutter app in the DOM, and its tests.
+///
+/// Consider this as the intersection in functionality between [html.ShadowRoot]
+/// (preferred Flutter rendering method) and [html.Document] (fallback).
+///
+/// Not to be confused with [html.DocumentOrShadowRoot].
+abstract class HostNode {
+  /// Retrieves the [html.Element] that currently has focus.
+  ///
+  /// See:
+  /// * [Document.activeElement](https://developer.mozilla.org/en-US/docs/Web/API/Document/activeElement)
+  html.Element? get activeElement;
+
+  /// Adds a node to the end of the child [nodes] list of this node.
+  ///
+  /// If the node already exists in this document, it will be removed from its
+  /// current parent node, then added to this node.
+  ///
+  /// This method is more efficient than `nodes.add`, and is the preferred
+  /// way of appending a child node.
+  ///
+  /// See:
+  /// * [Node.appendChild](https://developer.mozilla.org/en-US/docs/Web/API/Node/appendChild)
+  html.Node append(html.Node node);
+
+  /// Returns true if this node contains the specified node.
+  /// See:
+  /// * [Node.contains](https://developer.mozilla.org/en-US/docs/Web/API/Node.contains)
+  bool contains(html.Node? other);
+
+  /// Returns the currently wrapped [html.Node].
+  html.Node get node;
+
+  /// A modifiable list of this node's children.
+  List<html.Node> get nodes;
+
+  /// Finds the first descendant element of this document that matches the
+  /// specified group of selectors.
+  ///
+  /// [selectors] should be a string using CSS selector syntax.
+  ///
+  /// ```dart
+  /// var element1 = document.querySelector('.className');
+  /// var element2 = document.querySelector('#id');
+  /// ```
+  ///
+  /// For details about CSS selector syntax, see the
+  /// [CSS selector specification](http://www.w3.org/TR/css3-selectors/).
+  ///
+  /// See:
+  /// * [Document.querySelector](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector)
+  html.Element? querySelector(String selectors);
+
+  /// Finds all descendant elements of this document that match the specified
+  /// group of selectors.
+  ///
+  /// [selectors] should be a string using CSS selector syntax.
+  ///
+  /// ```dart
+  /// var items = document.querySelectorAll('.itemClassName');
+  /// ```
+  ///
+  /// For details about CSS selector syntax, see the
+  /// [CSS selector specification](http://www.w3.org/TR/css3-selectors/).
+  ///
+  /// See:
+  /// * [Document.querySelectorAll](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll)
+  List<html.Node> querySelectorAll(String selectors);
+}
+
+/// A partial [HostNode] implementation, backed by a [html.Node].
+///
+/// The implementation is partial (abstract) because Nodes alone can't implement
+/// the complete functionality.
+///
+/// The late property `_node` must be set in the extender's constructor.
+abstract class NodeHostNode implements HostNode {
```

Comment:
Can the partial functionality be included in `HostNode` itself? But also, it doesn't look like there's enough functionality to share. I'd even be OK if this is duplicated between `ElementHostNode` and `ShadowHostNode`. In fact, then we wouldn't need `_node` and `_element` in the same class (or `_node` and `_shadow`). Each class would have `_node` of the respective concrete type, Element or ShadowRoot.

Reply:
Yeah, I couldn't find a good way of initializing _node of the superclass, there's not a lot of "reused" code.